File: kotlin/src/main/kotlin/com/peshchuk/topjslibs/PageJsSources.kt
package com.peshchuk.topjslibs

private val SCRIPT_SRC = """<script[^>]+?src=["'](?:[^"']+/)?([^"']+?)["']""".toRegex()

class PageJsSources(private val fetchHtml: () -> String) {

    fun get(): Set<String> {
        val html = fetchHtml()
        val matchResults = SCRIPT_SRC.findAll(html)

        return matchResults
            .filter { matchResult -> matchResult.groups.size > 1 }
            .map { matchResult -> matchResult.groups[1]?.value!! }
            .toSet()
    }
}

File: scala/src/main/scala/com/peshchuk/topjslibs/PageJsSources.scala
package com.peshchuk.topjslibs

class PageJsSources(private val fetchHtml: () => String) {

  def get: Set[String] = {
    val html = fetchHtml()
    val matchData = PageJsSources.SCRIPT_SRC.findAllIn(html).matchData

    matchData
      .filter(_.groupCount > 0)
      .map(_.group(1))
      .toSet
  }
}

object PageJsSources {
  private final val SCRIPT_SRC = """<script[^>]+?src=["'](?:[^"']+/)?([^"']+?)["']""".r
}
